### PR TITLE
add service context map

### PR DESCRIPTION
**Pull request by qiwzhang**
to support multiple version of service config.

### CODE REVIEW

**Comment by mangchiandjjoe** [comment]
Are we working on the master branch for this feature?

**Comment by qiwzhang** [comment]
Yes

File: contrib/endpoints/include/api_manager/service_control.h [hunk 1]
```diff
@@ -45,6 +45,20 @@ struct Statistics {
 
   // Maximum report request size send to server.
   uint64_t max_report_size;
+
+  // operator+= to combine multiple statistics.
+  Statistics& operator+=(const Statistics& v) {
```
Comment (lizan) [inline contrib/endpoints/include/api_manager/service_control.h]: `Merge` instead of operator overloading?

Comment (qiwzhang) [inline contrib/endpoints/include/api_manager/service_control.h]: Done

File: contrib/endpoints/src/api_manager/api_manager_impl.cc [hunk 2]
```diff
@@ -27,26 +27,110 @@ ApiManagerImpl::ApiManagerImpl(std::unique_ptr<ApiManagerEnvInterface> env,
...
+}
+
+utils::Status ApiManagerImpl::GetStatistics(
+    ApiManagerStatistics *statistics) const {
+  memset(&statistics->service_control_statistics, 0,
+         sizeof(service_control::Statistics));
+  for (const auto &it : service_context_map_) {
+    if (it.second->service_control()) {
+      service_control::Statistics stat;
+      it.second->service_control()->GetStatistics(&stat);
```
Comment (lizan) [inline contrib/endpoints/src/api_manager/api_manager_impl.cc]: Handle error statuses returned from service_control?

Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/api_manager_impl.cc]: Done

File: contrib/endpoints/src/api_manager/api_manager_impl.cc [hunk 3]
```diff
@@ -27,26 +27,110 @@ ApiManagerImpl::ApiManagerImpl(std::unique_ptr<ApiManagerEnvInterface> env,
...
+    std::string config_id = config->service().id();
+    service_context_map_[config_id] = std::make_shared<context::ServiceContext>(
         global_context_, std::move(config));
+    // TODO: if this function is called at worker process, need to call
+    // service_context->service_control()->Init().
+    // ApiManagerImpl constructor is called at master process, not at worker
+    // process.
+    if (deploy_it) {
+      DeployConfigs({{config_id, 0}});
+    }
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/api_manager_impl.cc]: LoadServiceConfig at config.cc should have logged proper info if the service config is invalid.

File: contrib/endpoints/src/api_manager/api_manager_impl.cc [hunk 4]
```diff
@@ -27,26 +27,110 @@ ApiManagerImpl::ApiManagerImpl(std::unique_ptr<ApiManagerEnvInterface> env,
...
+    AddConfig(service_config, true);
   }
 
   check_workflow_ = std::unique_ptr<CheckWorkflow>(new CheckWorkflow);
   check_workflow_->RegisterAll();
 }
 
-void ApiManagerImpl::AddConfig(const std::string &service_config) {
+void ApiManagerImpl::AddConfig(const std::string &service_config,
+                               bool deploy_it) {
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/api_manager_impl.cc]: Good point, will update

Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/api_manager_impl.cc]: Done

**Pull request merged**